### PR TITLE
Update to Python3.5

### PR DESCRIPTION
**Pull request by ilan-gold**
This pull request contains the update to python3.5 outlined here: #2619

### CODE REVIEW

File: deployment/puppet/refinery/manifests/apache2.pp [hunk 1]
```diff
@@ -32,10 +32,11 @@
   }
 
   class { 'apache::mod::wsgi':
-    mod_path     => 'mod_wsgi.so',
-    package_name => 'libapache2-mod-wsgi',
+    mod_path         => 'mod_wsgi.so',
+    package_name     => 'libapache2-mod-wsgi-py3',
+    wsgi_python_home => "${virtualenv}",
+    wsgi_python_path => "${django_root}",
```
Comment (hackdna) [inline deployment/puppet/refinery/manifests/apache2.pp]: Are these options necessary? They already exist in `apache::vhost`

File: deployment/puppet/refinery/manifests/python.pp [hunk 2]
```diff
@@ -7,16 +7,21 @@
   $django_root         = $refinery::params::django_root,
 ) inherits refinery::params {
 
+
   class { '::python':
+    version    => 'python3.5',
```
Comment (hackdna) [inline deployment/puppet/refinery/manifests/python.pp:11]: Why not use the latest version 3.8?

File: deployment/puppet/refinery/manifests/apache2.pp [hunk 3]
```diff
@@ -132,7 +133,8 @@
       'refinery' => {
         'user'        => $app_user,
         'group'       => $app_group,
-        'python-path' => "${django_root}:${virtualenv}/lib/python2.7/site-packages",
+        'python-home' => "${virtualenv}",
+        "python-path" => "${django_root}"
```
Comment (hackdna) [inline deployment/puppet/refinery/manifests/apache2.pp]: It's best to use double quotes for interpolated strings and single quotes for regular strings. Also, add the trailing comma.
`"python-path" => "${django_root}",`
https://puppet.com/docs/puppet/5.5/style_guide.html

File: deployment/puppet/refinery/manifests/python.pp [hunk 4]
```diff
@@ -41,17 +46,20 @@
   }
   ->
   python::virtualenv { $virtualenv:
-    ensure  => present,
-    owner   => $app_user,
-    group   => $app_group,
-    require => [
+    version    => '3.5',
```
Comment (hackdna) [inline deployment/puppet/refinery/manifests/python.pp]: Is this not inherited from `::python` class?

File: deployment/puppet/refinery/manifests/python.pp [hunk 5]
```diff
@@ -41,17 +46,20 @@
...
+    ensure     => present,
+    owner      => $app_user,
+    group      => $app_group,
+    distribute => false,
```
Comment (hackdna) [inline deployment/puppet/refinery/manifests/python.pp]: Is this necessary?

File: deployment/puppet/refinery/manifests/python.pp [hunk 6]
```diff
@@ -72,12 +80,18 @@
...
   }
 
-  package { 'virtualenvwrapper': }
+  exec { 'get virtualenvwrapper':
+    command     => "python3 -m pip install virtualenvwrapper",
+    path        => [ '/bin/', '/sbin/' , '/usr/bin/', '/usr/sbin/' ],
+    user        => $app_user,
+    group       => $app_group,
+    require     => Python::Virtualenv[$virtualenv],
+  }
```
Comment (hackdna) [inline deployment/puppet/refinery/manifests/python.pp]: There should be no need for this since `virtualenvwrapper` package is available.

Btw, `python::pip` installs and manages packages from pip
https://forge.puppet.com/puppet/python/reference#pythonpip
In general, `exec` resource type should be used as a last resort

File: deployment/puppet/refinery/manifests/python.pp [hunk 7]
```diff
@@ -72,12 +80,18 @@
...
   ->
   file_line { "virtualenvwrapper_config":
-    path    => "/home/${app_user}/.profile",
-    line    => "source /etc/bash_completion.d/virtualenvwrapper",
-    require => Python::Virtualenv[$virtualenv],
+    path        => "/home/${app_user}/.profile",
+    line        => "export VIRTUALENVWRAPPER_PYTHON=/usr/bin/python3 && export PROJECT_HOME=/vagrant/refinery && export VIRTUALENVWRAPPER_VIRTUALENV=/usr/bin/virtualenv && export WORKON_HOME=/home/${app_user}/.virtualenv/ && source ./.local/bin/virtualenvwrapper.sh",
```
Comment (hackdna) [inline deployment/puppet/refinery/manifests/python.pp]: This should suffice:
```shell
VIRTUALENVWRAPPER_PYTHON='/usr/bin/python3'
source /usr/local/bin/virtualenvwrapper.sh
```
https://askubuntu.com/questions/974511/ubuntu-16-04-virtualenvwrapper-python3-6-path

File: refinery/analysis_manager/models.py [hunk 8]
```diff
@@ -43,15 +43,15 @@ class AnalysisStatus(models.Model):
...
+    def __str__(self):
         return self.analysis.name
 
     def set_galaxy_history_state(self, state):
         """
         Set the `galaxy_history_state` of an analysis
         :param state: a valid GALAXY_HISTORY_STATE
         """
-        if state in dict(self.GALAXY_HISTORY_STATES).keys():
+        if state in list(dict(self.GALAXY_HISTORY_STATES).keys()):
```
Comment (hackdna) [inline refinery/analysis_manager/models.py]: `list()` should not be necessary since `keys()` returns an iterable

File: refinery/analysis_manager/models.py [hunk 9]
```diff
@@ -62,7 +62,7 @@ def set_galaxy_import_state(self, state):
         Set the `galaxy_import_state` of an analysis
         :param state: a valid GALAXY_HISTORY_STATE
         """
-        if state in dict(self.GALAXY_HISTORY_STATES).keys():
+        if state in list(dict(self.GALAXY_HISTORY_STATES).keys()):
```
Comment (hackdna) [inline refinery/analysis_manager/models.py]: `list()` should not be necessary since `keys()` returns an iterable

File: deployment/puppet/refinery/manifests/python.pp (hunk 2 above)
Comment (ilan-gold) [reply deployment/puppet/refinery/manifests/python.pp:11]: the ubuntu version we're on comes with 3.5 - otherwise we'd have to pull from [here](https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa).  also, i don't know if django 1.8 supports 3.8 officially - the website was somewhat unclear.

File: deployment/puppet/refinery/manifests/python.pp (hunk 6 above)
Comment (ilan-gold) [reply deployment/puppet/refinery/manifests/python.pp]: what do you mean that it is "available?"

File: deployment/puppet/refinery/manifests/python.pp (hunk 4 above)
Comment (ilan-gold) [reply deployment/puppet/refinery/manifests/python.pp]: related to the above, if we were to use a different version of python from deadsnakes, it was not my experience that this would inherit.

Comment (ilan-gold) [reply deployment/puppet/refinery/manifests/python.pp]: thus i left it alone since it could be helpful to have around in the future and i prefer the specificity

File: refinery/analysis_manager/models.py (hunk 8 above)
Comment (ilan-gold) [reply refinery/analysis_manager/models.py]: i am aware - this was from 2to3, whose changes i did not touch.  i can update these if you'd prefer though

File: deployment/puppet/refinery/manifests/python.pp (hunk 2 above)
Comment (hackdna) [reply deployment/puppet/refinery/manifests/python.pp:11]: Good point, Django 1.8 - 1.10 support only up to 3.5 (https://docs.djangoproject.com/en/1.11/faq/install/#what-python-version-can-i-use-with-django)

However, 3.5 end-of-life is 2020-09-13 (https://devguide.python.org/#status-of-python-branches) - that should motivate upgrade to Ubuntu 18.04 (it comes with 3.6) and to Django 1.11+

Comment (ilan-gold) [reply deployment/puppet/refinery/manifests/python.pp:11]: yup, i filed that away in the back of my head while doing this :)

File: deployment/puppet/refinery/manifests/apache2.pp (hunk 1 above)
Comment (ilan-gold) [reply deployment/puppet/refinery/manifests/apache2.pp]: i will look into this - it might be touching a different .conf file than the apache::vhost but i can't remember.  i've looked at the puppet source code for this ( while going down the rabbit hole that led me to change these similar settings in apache::vhost) and it's very opaque so i would not be surprised if it were necessary.  also this is another thing to watch out for when we  upgrade to python3.8.  i don't think there a version of libapache2-mod-wsgi-py3 out for python3.8 so we'll need to switch over to pip-installing it - i think the writer of the package suggests doing it this way going forward anyway which might explain its absence.  in any case, the way to do this was not simple to figure out, but i know how now :)

File: deployment/puppet/refinery/manifests/python.pp (hunk 6 above)
Comment (ilan-gold) [reply deployment/puppet/refinery/manifests/python.pp]: in short, though, this is a rabbit hole i chose not to go down any further.  i think this is a known issue with pip and python and ubuntu: https://askubuntu.com/a/1106328

File: deployment/puppet/refinery/manifests/python.pp (hunk 7 above)
Comment (ilan-gold) [reply deployment/puppet/refinery/manifests/python.pp]: we need the `WORKON_HOME` as well but otherwise we got rid of the other fields

File: deployment/puppet/refinery/manifests/apache2.pp (hunk 1 above)
Comment (ilan-gold) [reply deployment/puppet/refinery/manifests/apache2.pp]: i have removed these now - i appear to have forgotten reading this: https://stackoverflow.com/a/50165569

essentially, we have no choice but to download mod_wsgi from PyPi unless we want to have a broken version and nothing works

File: deployment/puppet/refinery/manifests/python.pp (hunk 6 above)
Comment (hackdna) [reply deployment/puppet/refinery/manifests/python.pp]: `virtualenvwrapper` is available as an APT package

File: refinery/analysis_manager/tasks.py [hunk 10]
```diff
@@ -475,7 +478,7 @@ def _get_galaxy_download_task_ids(analysis):
     except galaxy.client.ConnectionError as exc:
         error_msg = \
             "Error downloading Galaxy history files for analysis '%s': %s"
-        logger.error(error_msg, analysis.name, exc.message)
+        logger.error(error_msg, analysis.name, str(exc))
```
Comment (hackdna) [inline refinery/analysis_manager/tasks.py]: I don't think explicit conversion to string is required

File: deployment/puppet/refinery/manifests/python.pp [hunk 11]
```diff
@@ -7,16 +7,21 @@
   $django_root         = $refinery::params::django_root,
 ) inherits refinery::params {
 
+
```
Comment (hackdna) [inline deployment/puppet/refinery/manifests/python.pp]: Unnecessary blank line

File: deployment/puppet/refinery/manifests/python.pp [hunk 12]
```diff
@@ -7,16 +7,21 @@
...
   class { '::python':
+    version    => 'python3.5',
+    ensure     => 'latest',
     dev        => 'present',
     gunicorn   => 'absent',
     pip        => 'latest',
-    virtualenv => 'present',
+    virtualenv => 'present'
```
Comment (hackdna) [inline deployment/puppet/refinery/manifests/python.pp]: Need a trailing comma

File: .travis.yml [hunk 13]
```diff
@@ -17,6 +17,9 @@ addons:
       - postgresql-client-10
       - libgconf-2-4  # for Cypress
 
+before_install:
+  - sudo apt-get install apache2-dev
```
Comment (hackdna) [inline .travis.yml]: What is this for?

Comment (ilan-gold) [reply .travis.yml]: mod_wsgi: https://modwsgi.readthedocs.io/en/develop/user-guides/quick-installation-guide.html

Comment (ilan-gold) [reply .travis.yml]: (since we install it via pip now and thus it gets installed on travis too)

File: requirements.txt [hunk 14]
```diff
@@ -28,16 +29,19 @@ flower==0.9.1
 jsonschema==2.6.0
 Markdown==2.2.1
 mock==2.0.0
-mockcache==1.0.2
+mod_wsgi==4.6.7
```
Comment (hackdna) [inline requirements.txt]: `mod_wsgi` should be installed using `apt` as before

File: requirements.txt [hunk 15]
```diff
@@ -6,9 +6,10 @@ cffi==1.9.1
 codecov==2.0.11
 coverage==4.0.3
 cryptography==2.5
+Cython==0.29.13
```
Comment (hackdna) [inline requirements.txt]: Why is this needed?

File: requirements.txt [hunk 16]
```diff
@@ -28,16 +29,19 @@ flower==0.9.1
 jsonschema==2.6.0
 Markdown==2.2.1
 mock==2.0.0
-mockcache==1.0.2
+mod_wsgi==4.6.7
+git+git://github.com/lunant/mockcache@master
+ndg-httpsclient==0.3.2
 networkx==1.7
 psycopg2-binary==2.7.4
+pyasn1==0.1.9
```
Comment (hackdna) [inline requirements.txt]: Why is this required?

File: requirements.txt [hunk 17]
```diff
@@ -28,16 +29,19 @@ flower==0.9.1
 jsonschema==2.6.0
 Markdown==2.2.1
 mock==2.0.0
-mockcache==1.0.2
+mod_wsgi==4.6.7
+git+git://github.com/lunant/mockcache@master
+ndg-httpsclient==0.3.2
```
Comment (hackdna) [inline requirements.txt]: Why is this required?

File: requirements.txt [hunk 18]
```diff
@@ -28,16 +29,19 @@ flower==0.9.1
 jsonschema==2.6.0
 Markdown==2.2.1
 mock==2.0.0
-mockcache==1.0.2
+mod_wsgi==4.6.7
+git+git://github.com/lunant/mockcache@master
```
Comment (hackdna) [inline requirements.txt:31]: The last commit to that repo was in 2016. I looks like this module is no longer maintained, so need to find a replacement or remove it.

Comment (ilan-gold) [reply requirements.txt:31]: Nils and I both felt comfortable pinning it to master (we can also do a specific commit) but removing it should be a separate issue independent of this pull request if you feel strongly.

File: deployment/puppet/refinery/manifests/python.pp (hunk 4 above)
Comment (hackdna) [reply deployment/puppet/refinery/manifests/python.pp]: The problem is that this creates two sources of truth for Python version which can lead to errors. We are using the OS-provided Python now, so this is not needed.

Comment (ilan-gold) [reply deployment/puppet/refinery/manifests/python.pp]: That sounds right, good catch - I hadn't thought that putting that there with the OS version would download an extra version like it does for a different number with deadsnakes i.e 3.7

File: requirements.txt (hunk 18 above)
Comment (hackdna) [reply requirements.txt:31]: Does Nils know that this package is no longer maintained?

File: deployment/puppet/refinery/manifests/python.pp [hunk 19]
```diff
@@ -72,12 +79,18 @@
...
+    group       => $app_group,
+    require     => Python::Virtualenv[$virtualenv],
+  }
   ->
   file_line { "virtualenvwrapper_config":
-    path    => "/home/${app_user}/.profile",
-    line    => "source /etc/bash_completion.d/virtualenvwrapper",
-    require => Python::Virtualenv[$virtualenv],
+    path        => "/home/${app_user}/.profile",
+    line        => "export VIRTUALENVWRAPPER_PYTHON=/usr/bin/python3 && export WORKON_HOME=/home/${app_user}/.virtualenvs/ && source ./.local/bin/virtualenvwrapper.sh",
```
Comment (hackdna) [inline deployment/puppet/refinery/manifests/python.pp]: I don't think this line should be changed at all if virtualenvwrapper is installed by APT.

File: refinery/analysis_manager/tasks.py [hunk 20]
```diff
@@ -3,7 +3,7 @@
 
 @author: nils
 '''
-import urlparse
+import urllib.parse
```
Comment (hackdna) [inline refinery/analysis_manager/tasks.py]: just `import urllib` is sufficient

File: refinery/core/models.py [hunk 21]
```diff
@@ -899,15 +899,15 @@ def get_node_collection(self):
                 NodeCollection.MultipleObjectsReturned) as exc:
             logger.error('Failed to get NodeCollection for '
                          'Investigation %s: %s',
-                         unicode(self.investigation), exc)
+                         str(self.investigation), exc)
```
Comment (hackdna) [inline refinery/core/models.py]: Explicit cast to string is probably unnecessary. Also, this whole statement can be compacted to two lines with sacrificing readability.

File: refinery/core/models.py [hunk 22]
```diff
@@ -2120,9 +2118,9 @@ def get_csv_row(self, aggregates=False):
         def get_aggregate_sum(field_name):
             if not aggregates:
                 return getattr(self, field_name)
-            return SiteStatistics.objects.filter(
+            return list(SiteStatistics.objects.filter(
```
Comment (hackdna) [inline refinery/core/models.py:2124]: Probably unnecessary

File: refinery/core/views.py [hunk 23]
```diff
@@ -2,7 +2,9 @@
 from datetime import timedelta
 import json
 import logging
-import urllib
+import urllib.request
+import urllib.parse
+import urllib.error
```
Comment (hackdna) [inline refinery/core/views.py]: `import urllib` should be sufficient

File: refinery/data_set_manager/isa_tab_parser.py [hunk 24]
```diff
@@ -1173,7 +1168,7 @@ def get_dataset_name(self, path):
         investigation_title = True
         title = None
 
-        with open(investigation_file_name, 'r') as f:
+        with open(investigation_file_name, 'r', encoding='latin-1') as f:
```
Comment (hackdna) [inline refinery/data_set_manager/isa_tab_parser.py:1171]: Why 'latin-1'?

File: refinery/data_set_manager/test_models.py [hunk 25]
```diff
@@ -216,48 +216,54 @@ def test_get_assay_count(self):
...
-             u'rfc134.txt', u'rfc174.txt', u'rfc177.txt', u'rfc178.txt',
-             u'rfc86.txt', u'rfc94.txt']
+            ['rfc-test.zip', 'rfc111.txt', 'rfc125.txt', 'rfc126.txt',
+             'rfc134.txt', 'rfc174.txt', 'rfc177.txt', 'rfc178.txt',
+             'rfc86.txt', 'rfc94.txt']
         )
 
     def test_get_datafile_names_local_only(self):
-        with open(os.path.join(TEST_DATA_BASE_PATH, "rfc-test.zip")) as isatab:
+        with open(os.path.join(TEST_DATA_BASE_PATH, "rfc-test.zip"), 'rb') \
```
Comment (hackdna) [inline refinery/data_set_manager/test_models.py]: may be:
```python
isa_archive_path = os.path.join(TEST_DATA_BASE_PATH, 'rfc-test.zip')
```

File: refinery/data_set_manager/test_utils.py [hunk 26]
```diff
@@ -448,11 +449,11 @@ def test_is_field_in_hidden_list(self):
     def test_generate_solr_params_no_params_returns_obj(self):
         # empty params
         query = generate_solr_params_for_assay(QueryDict({}), self.valid_uuid)
-        self.assertItemsEqual(sorted(query.keys()), ['json', 'params'])
+        self.assertCountEqual(sorted(query.keys()), ['json', 'params'])
```
Comment (hackdna) [inline refinery/data_set_manager/test_utils.py:452]: This looks like a much weaker test. Is there anything similar to `assertItemsEqual`?

File: refinery/data_set_manager/views.py [hunk 27]
```diff
@@ -367,7 +369,11 @@ def post(self, request, *args, **kwargs):
                 try:
                     response = import_by_file(f)
                 except Exception as e:
-                    logger.error(traceback.format_exc(e))
+                    etype, value, tb = sys.exc_info()
+                    logger.error("{} {}".format(
+                        PARSER_UNEXPECTED_ERROR_MESSAGE,
+                        "".join(traceback.format_exception(etype, value, tb))
+                    ))
```
Comment (hackdna) [inline refinery/data_set_manager/views.py:376]: Why not just `logger.error(e)`?

File: refinery/data_set_manager/test_utils.py (hunk 26 above)
Comment (ilan-gold) [reply refinery/data_set_manager/test_utils.py:452]: my understanding is that they are identical - i have a link to a reference in the issue for this but here is another: https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertItemsEqual

File: refinery/data_set_manager/isa_tab_parser.py (hunk 24 above)
Comment (ilan-gold) [reply refinery/data_set_manager/isa_tab_parser.py:1171]: i think this is rectifying another python hiccup: http://python-notes.curiousefficiency.org/en/latest/python3/text_file_processing.html.  seems to best to just leave it as close to 2 as possible

File: refinery/ui/cypress/support/commands.js [hunk 28]
```diff
@@ -36,7 +36,7 @@ function shell_quote(str) {
 
 Cypress.Commands.add('wrap_cmd',
   function(manage_cmd) {
-    var cd_cmd = "cd .. && " + manage_cmd;
+    var cd_cmd = "cd .. && source ~/virtualenv/python3.5.6/bin/activate && " + manage_cmd;
```
Comment (hackdna) [inline refinery/ui/cypress/support/commands.js:39]: Appears to be redundant with `workon refinery-platform`

Comment (ilan-gold) [reply refinery/ui/cypress/support/commands.js:39]: this is specifically for travis - the `workon` command is not run because `vagrant` is not run

File: refinery/analysis_manager/tasks.py (hunk 20 above)
Comment (ilan-gold) [reply refinery/analysis_manager/tasks.py]: i tried this:
```import urllib
urllib.parse('foo')
Traceback (most recent call last):
  File "<input>", line 1, in <module>
TypeError: 'module' object is not callable
```

i also don't want to `from urllib import parse` (or its ilk in `urllib` since those words are so common)

File: requirements.txt (hunk 18 above)
Comment (ilan-gold) [reply requirements.txt:31]: yes - we only use this in our unit testing

File: refinery/ui/cypress/support/commands.js (hunk 28 above)
Comment (hackdna) [reply refinery/ui/cypress/support/commands.js:39]: It was working previously, so why wouldn't `workon` run now? What is the connection between virtualenvwrapper and Vagrant exactly?

Comment (ilan-gold) [reply refinery/ui/cypress/support/commands.js:39]: The `workon_cmd` is not run in travis, but the `cd_cmd` is.  I have not tried it for a bit, and so I could be wrong, but it appeared that the virtualenv was not persisting to the cypress test for some reason.  Thus it needed to be reactivated.

File: requirements.txt (hunk 15 above)
Comment (ilan-gold) [reply requirements.txt]: `pysam` doesn't pin its `Cython` dependency well so I figured it safest for us to pin ours (I did have some trouble with incorrect `Cython` versions locally but didn't spend too much time digging into this)

Comment (hackdna) [reply requirements.txt]: Pinning 3rd party dependencies is problematic because it can cause resolution conflicts with other packages and also, it is easy to forget to remove them if the parent package is removed.

File: refinery/data_set_manager/views.py (hunk 27 above)
Comment (ilan-gold) [reply refinery/data_set_manager/views.py:376]: I just tried to replicate the behavior - I did not try to change what it did.

File: refinery/analysis_manager/tasks.py [hunk 29]
```diff
@@ -3,7 +3,7 @@
 
 @author: nils
 '''
-import urlparse
+import urllib
```
Comment (hackdna) [inline refinery/analysis_manager/tasks.py]: It looks like this kind of import is not going to work after all in this case, so I'd do `from urllib.parse import urljoin`, etc.

File: refinery/data_set_manager/utils.py [hunk 30]
```diff
@@ -951,13 +950,13 @@ def get_file_url_from_node_uuid(node_uuid, require_valid_url=False):
             return core.utils.build_absolute_url(url) if url else None
         except ValueError as e:
             logger.error('URL {} is not a valid relative url'.format(str(url)))
-            raise type(e)(e.message)
+            raise type(e)(str(e))
```
Comment (hackdna) [inline refinery/data_set_manager/utils.py]: Simply `raise`?

File: refinery/data_set_manager/utils.py [hunk 31]
```diff
@@ -951,13 +950,13 @@ def get_file_url_from_node_uuid(node_uuid, require_valid_url=False):
...
         except RuntimeError as e:
             logger.error('Could not build absolute URL for {}'.format(
                     str(url)
                 )
             )
-            raise type(e)(e.message)
+            raise type(e)(str(e))
```
Comment (hackdna) [inline refinery/data_set_manager/utils.py]: Simply `raise`?

File: refinery/data_set_manager/utils.py (hunk 30 above)
Comment (ilan-gold) [reply refinery/data_set_manager/utils.py]: I wanted to keep the message

File: refinery/data_set_manager/utils.py (hunk 31 above)
Comment (ilan-gold) [reply refinery/data_set_manager/utils.py]: I wanted to keep the message

File: refinery/file_store/tasks.py [hunk 32]
```diff
@@ -1,10 +1,9 @@
-from __future__ import division
+
 import contextlib
 import os
 import tempfile
 import threading
-import urllib2
-import urlparse
+import urllib
```
Comment (hackdna) [inline refinery/file_store/tasks.py]:
```python
from urllib.parse import urlparse
from urllib.request import urlopen
```

File: refinery/file_store/tasks.py [hunk 33]
```diff
@@ -234,16 +233,16 @@ def import_url_to_path(self, source_url, target_name=None):
         storage = SymlinkedFileSystemStorage()
         # remove query string from URL before extracting file name
         if target_name is None:
-            target_name = os.path.basename(urlparse.urlparse(source_url).path)
+            target_name = os.path.basename(urllib.parse(source_url).path)
```
Comment (hackdna) [inline refinery/file_store/tasks.py]: `urlparse`

File: refinery/file_store/tasks.py [hunk 34]
```diff
@@ -234,16 +233,16 @@ def import_url_to_path(self, source_url, target_name=None):
...
         file_store_name = storage.get_name(target_name)
         file_store_path = storage.path(file_store_name)
 
         logger.debug("Transferring from '%s' to '%s'",
                      source_url, file_store_path)
         make_dir(os.path.dirname(file_store_path))
         try:
-            with contextlib.closing(urllib2.urlopen(source_url,
-                                                    timeout=30)) as response, \
+            with contextlib.closing(urllib.request.urlopen(source_url,
```
Comment (hackdna) [inline refinery/file_store/tasks.py]: `urlopen`

File: refinery/file_store/tasks.py [hunk 35]
```diff
@@ -262,14 +261,15 @@ def import_url_to_s3(self, source_url, target_name=None):
         storage = S3MediaStorage()
         # remove query string from URL before extracting file name
         if target_name is None:
-            target_name = os.path.basename(urlparse.urlparse(source_url).path)
+            target_name = os.path.basename(urllib.parse(source_url).path)
```
Comment (hackdna) [inline refinery/file_store/tasks.py]: `urlparse`

File: refinery/file_store/tasks.py [hunk 36]
```diff
@@ -262,14 +261,15 @@ def import_url_to_s3(self, source_url, target_name=None):
...
         file_store_name = storage.get_name(target_name)
 
         logger.debug("Transferring from '%s' to 's3://%s/%s'",
                      source_url, settings.MEDIA_BUCKET, file_store_name)
         try:
-            with contextlib.closing(urllib2.urlopen(source_url, timeout=30)) \
-                    as response:
+            with contextlib.closing(
+                    urllib.request.urlopen(source_url, timeout=30)
```
Comment (hackdna) [inline refinery/file_store/tasks.py]: `urlopen`

File: refinery/file_store/utils.py [hunk 37]
```diff
@@ -4,9 +4,7 @@
 import shutil
 import stat
 import hashlib
-import urllib2
-import urlparse
-
+import urllib
```
Comment (hackdna) [inline refinery/file_store/utils.py]:
```python
from urllib.request import urlopen
from urllib.parse import urlparse
```

File: refinery/file_store/utils.py [hunk 38]
```diff
@@ -158,9 +156,9 @@ def get_file_size(file_location):
             return UNKNOWN_FILE_SIZE
     else:
         try:
-            with contextlib.closing(urllib2.urlopen(file_location,
-                                                    timeout=30)) as response:
-                return int(response.info().getheader('Content-Length'))
+            with contextlib.closing(urllib.request.urlopen(file_location,
```
Comment (hackdna) [inline refinery/file_store/utils.py]: `urlopen`
Also, `contextlib.closing` is no longer necessary: https://docs.python.org/3/howto/urllib2.html#fetching-urls

File: refinery/file_store/utils.py [hunk 39]
```diff
@@ -158,9 +156,9 @@ def get_file_size(file_location):
...
+                                    timeout=30)) as response:
+                return int(response.info().get('Content-Length'))
```
Comment (hackdna) [inline refinery/file_store/utils.py]: Should this be: `return int(response.getheader('Content-Length'))`?
https://docs.python.org/3/library/http.client.html#http.client.HTTPResponse.getheader

File: refinery/file_store/utils.py [hunk 40]
```diff
@@ -201,7 +199,7 @@ def parse_s3_url(url):
     """Return a tuple containing S3 bucket name and object key given S3
     protocol URL (s3://bucket-name/key)
     """
-    result = urlparse.urlparse(url)
+    result = urllib.parse.urlparse(url)
```
Comment (hackdna) [inline refinery/file_store/utils.py]: `urlparse`

File: refinery/tool_manager/test_views.py [hunk 41]
```diff
@@ -600,7 +602,7 @@ def test_tool_creation_with_same_display_name_yields_helpful_error(self):
         self.assertEqual(
             "A Tool already exists with a display_name of: '{}'".format(
                 display_name
-            ),
+            ).encode(),
             self.post_response.content
```
Comment (hackdna) [inline refinery/tool_manager/test_views.py:606]: `self.post_response.content.decode()` instead?

File: refinery/ui/cypress/support/commands.js (hunk 28 above)
Comment (hackdna) [reply refinery/ui/cypress/support/commands.js:39]: It is still not clear why did this addition become necessary. What was the error message? Also, that path does not exist in Vagrant:
```shell
vagrant@refinery:~$ ls ~/virtualenv/python3.5.6/bin/activate
ls: cannot access '/home/vagrant/virtualenv/python3.5.6/bin/activate': No such file or directory
```

File: refinery/data_set_manager/utils.py (hunk 31 above)
Comment (hackdna) [reply refinery/data_set_manager/utils.py]: The message is kept as part of the exception that is re-raised:
```python
>>> try:
...     raise RuntimeError('blah')
... except RuntimeError:
...     raise
...
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
RuntimeError: blah
```

File: refinery/core/models.py (hunk 22 above)
Comment (ilan-gold) [reply refinery/core/models.py:2124]: this is necessary - you cannot index a `dict_value` object

Comment (hackdna) [reply refinery/core/models.py:2124]: and if you need a list just remove `[0]`:
```python
>>> SiteStatistics.objects.all().aggregate(Sum('datasets_shared')).values()
[230]
```

File: refinery/ui/cypress/support/commands.js (hunk 28 above)
Comment (ilan-gold) [reply refinery/ui/cypress/support/commands.js:39]: The build fails in `travis` without it, if my memory serves.  I'll checkout a separate branch to confirm and remove the line to see  what happens.  If it is necessary, I'll move it so that it executes separate from the other commands.

File: refinery/analysis_manager/tasks.py (hunk 29 above)
Comment (ilan-gold) [reply refinery/analysis_manager/tasks.py]: Can you be more specific? This works in a python shell.

Comment (hackdna) [reply refinery/analysis_manager/tasks.py]: No, it doesn't:
```python
>>> import urllib
>>> urllib.parse.urlparse('foo')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'urllib' has no attribute 'parse'
>>> from urllib.parse import urlparse
>>> urlparse('foo')
ParseResult(scheme='', netloc='', path='foo', params='', query='', fragment='')
```

File: refinery/tool_manager/test_views.py (hunk 41 above)
Comment (ilan-gold) [reply refinery/tool_manager/test_views.py:606]: I thought it better to leave the `content` untouched.

File: refinery/core/models.py (hunk 22 above)
Comment (ilan-gold) [reply refinery/core/models.py:2124]: That returns a dictionary - I think we need the value of the dictionary, specifically the first one to return the proper element.
